packages/serve/src/startDevServer.ts

```diff
@@ -11,21 +11,25 @@ import Server from 'webpack-dev-server/lib/Server';
  */
 export default function startDevServer(compiler, options): void {
     const firstWpOpt = compiler.compilers ? compiler.compilers[0].options : compiler.options;
-    const devServerOptions = firstWpOpt.devServer || {};
+    let devServerOptions = firstWpOpt.devServer || {};
+
+    // set default host and port
+    devServerOptions = { host: 'localhost', port: 8080, ...devServerOptions };
 
-    const host = options.host || devServerOptions.host || 'localhost';
-    const port = options.port || devServerOptions.port || 8080;
     // socket should not have a default value, because it should only be used if the
     // user explicitly provides it
     const socket = options.socket || devServerOptions.socket;
 
-    options.host = host;
-    options.port = port;
     if (socket) {
         options.socket = socket;
     }
 
-    const server = new Server(compiler, options);
+    // Compose config from the CLI and devServer object from webpack config
+    const serverConfig = { ...options, ...devServerOptions };
```

Comment:
I think I understand what you are saying. Do you mean the issue is that if a CLI flag has a default value, it will override `devServer` regardless of what the user has set for that value in `devServer`? Yes, this is an issue.

Our goal is that CLI args should take precedence over `devServer`, so this line here is not the perfect solution.

What we need to do is find a way to check if the user has actually provided the CLI flag, or if it is just getting passed in as a default value. It might be easier to hold off on this until we swap over to `commander` as the CLI flags parser. Ideally we can pass a CLI options object into `startDevServer` which indicates whether each flag was defaulted to, or whether the user explicitly provided the flag.

Alternatively, we should just not have dev server CLI flag default values.

Reply:
> I think I understand what you are saying. Do you mean the issue is that if a CLI flag has a default value, it will override devServer regardless of what the user has set for that value in devServer? Yes, this is an issue.

Precisely.

That makes sense to me, lets revisit this after we're on the new parser.

Reply:
Can you find a case when we don't want the devServer config to override? I could think of `--hot` flag, when passed will appear in the CLI options object, and if `devServer` contains `hot: false` then the CLI option won't override this way, but nobody does will explicitly pass `hot: false` in their config right, since it's default?

Reply:
/cc @Loonride how do we proceed here?

Reply:
I think the goal may be to just make dev server flags not set default values. If we do that, this shouldn't be an issue. But in that case it would still be nice for users to see what values are defaulted to later on when they do like `webpack-cli help serve`. Will think about this a bit and come back to it

Reply:
Did you get any findings about this @Loonride ?

Reply:
Sorry for the delay, I can look at this more closely now that we are close to webpack-dev-server v4

Reply:
Awesome, let me know your thoughts :)